crates/ruff_linter/src/rules/ruff/rules/spurious_async.rs

```diff
@@ -0,0 +1,83 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::visitor;
+use ruff_python_ast::visitor::Visitor;
+use ruff_python_ast::{self as ast, Expr, Stmt};
+use ruff_text_size::TextRange;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for functions declared `async` that do not yield or otherwise use features requiring the
+/// function to be declared `async`.
+///
+/// ## Why is this bad?
+/// Declaring a function `async` when it's not is usually a mistake, and may artificially limit the
+/// contexts where that function may be called.
+#[violation]
+pub struct SpuriousAsync {
+    name: String,
+}
+
+impl Violation for SpuriousAsync {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let SpuriousAsync { name } = self;
+        format!("Function `{name}` is declared `async`, but doesn't yield or use async features.")
+    }
+}
+
+#[derive(Default)]
+struct YieldingExprVisitor {
+    found_yield: bool,
+}
+
+impl<'a> Visitor<'a> for YieldingExprVisitor {
+    fn visit_expr(&mut self, expr: &'a Expr) {
+        match expr {
+            Expr::Await(_) => {
+                self.found_yield = true;
+            }
+            _ => visitor::walk_expr(self, expr),
+        }
+    }
+    fn visit_stmt(&mut self, stmt: &'a Stmt) {
+        match stmt {
+            Stmt::With(ast::StmtWith { is_async: true, .. }) => {
+                self.found_yield = true;
+            }
+            Stmt::For(ast::StmtFor { is_async: true, .. }) => {
+                self.found_yield = true;
+            }
+            _ => visitor::walk_stmt(self, stmt),
+        }
+    }
+}
+
+/// RUF029
+pub(crate) fn spurious_async(
+    checker: &mut Checker,
+    is_async: bool,
+    name: &str,
+    body: &[Stmt],
+    range: TextRange,
+) {
+    if !is_async {
+        return;
+    }
+
+    let yields = {
+        let mut visitor = YieldingExprVisitor { found_yield: false };
```

Comment:
Nit: You could derive `Default` for `YieldingExprVisitor` so that this becomes
```suggestion
        let mut visitor = YieldingExprVisitor::default();
```

Reply:
Yes, I followed that pattern in an earlier draft when the field was a vector of yielding-expression-TextRanges, but reverted to explicit construction when I changed it to a boolean. Is the convention to always prefer default where applicable?

Reply:
There's no strong convention for private types. For public types, yes, I would say we prefer using `::default` or `::new` to encapsulate the initialization logic. Here I would do it because it's less to type but that's also why it is a Nit. It's up to you what you prefer.

Reply:
I've made this change below.